include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,366 @@
+    friend ::std::basic_istream<CharT, Traits>& 
+    operator>>(::std::basic_istream<CharT, Traits>&, philox_engine<UIntType_, w_, n_, r_, consts_...>&);
+
+private:
+    /* Internal generation Philox kernel */
+    void generate() {
+        if constexpr (word_count == 2) {
+                scalar_type R0 = (state_.X[0]) & in_mask;
+                scalar_type L0 = (state_.X[1]) & in_mask;
+                scalar_type K0 = (state_.K[0]) & in_mask;
+                for (size_t i = 0; i < round_count; ++i) {
+                    auto [hi0, lo0] = detail::mulhilo<scalar_type, word_size>(R0, multipliers[0]);
+                    R0 = hi0 ^ K0 ^ L0;
+                    L0 = lo0;
+                    K0 = (K0 + round_consts[0]) & in_mask;
+                }
+                state_.Y[0] = R0;
+                state_.Y[1] = L0;
+        }
+        else if constexpr (word_count == 4) {
+                scalar_type R0 = (state_.X[0]) & in_mask;
+                scalar_type L0 = (state_.X[1]) & in_mask;
+                scalar_type R1 = (state_.X[2]) & in_mask;
+                scalar_type L1 = (state_.X[3]) & in_mask;
+                scalar_type K0 = (state_.K[0]) & in_mask;
+                scalar_type K1 = (state_.K[1]) & in_mask;
+                for (size_t i = 0; i < round_count; ++i) {
+                    auto [hi0, lo0] = detail::mulhilo<scalar_type, word_size>(R0, multipliers[0]);
+                    auto [hi1, lo1] = detail::mulhilo<scalar_type, word_size>(R1, multipliers[1]);
+                    R0 = hi1 ^ L0 ^ K0;
+                    L0 = lo1;
+                    R1 = hi0 ^ L1 ^ K1; 
+                    L1 = lo0; 
+                    K0 = (K0 + round_consts[0]) & in_mask;
+                    K1 = (K1 + round_consts[1]) & in_mask;
+                }
+                state_.Y[0] = R0;
+                state_.Y[1] = L0;
+                state_.Y[2] = R1;
+                state_.Y[3] = L1;
+        }
+    }
+};
+
+template<class CharT, class Traits, typename UIntType, ::std::size_t w, ::std::size_t n, ::std::size_t r, UIntType... consts>
+::std::basic_ostream<CharT, Traits>& 
+operator<<(::std::basic_ostream<CharT, Traits>& os, const philox_engine<UIntType, w, n, r, consts...>& engine) {
+    internal::save_stream_flags<CharT, Traits> __flags(os);
+
+    os.setf(::std::ios_base::dec | ::std::ios_base::left);
+    CharT sp = os.widen(' ');
+    os.fill(sp);
+
+    for(auto x_elm: engine.state_.X) {
+        os << x_elm << sp;
+    }
+    for(auto k_elm: engine.state_.K) {
+        os << k_elm << sp;
+    }
+    for(auto y_elm: engine.state_.Y) {
+        os << y_elm << sp;
+    }
+    os << engine.state_.idx;
+    
+    return os;
+}
+
+template<typename UIntType, ::std::size_t w, ::std::size_t n, ::std::size_t r, UIntType... consts>
+const sycl::stream&
+operator<<(const sycl::stream& os, const philox_engine<UIntType, w, n, r, consts...>& engine) {
+    for(auto x_elm: engine.state_.X) {
```

Comment:
should we set:

internal::save_stream_flags<CharT, Traits> __flags(os);
  os.setf(::std::ios_base::dec | ::std::ios_base::left);
  CharT sp = os.widen(' ');
  os.fill(sp);
  
  here as well? and ' ' -> sp in lines 321, 324, 327

Reply:
I'm not sure. I did `operator<<` for `sycl::stream` similar to Kate's changes here #1636. All API for `sycl::stream` don't have the format set-up at the beginning(probably, it's not applicable for `sycl::stream`?)

Reply:
Discussed offline with @aelizaro for sycl::stream not applicable